File: TypeScript/ExampleTests/common/FrameValuesTyped.ts
import { FrameValuesString } from "./FrameValuesString.js";
import * as _json from "./json.js";

export class FrameValuesTyped {
  frame: number;
  roll1: string;
  roll2: string;
  roll3: string;
  score: string;
  totalScore: string;

  constructor(frame: number, roll1: string, roll2: string, roll3: string, score: string, totalScore: string) {
    this.frame = frame;
    this.roll1 = roll1;
    this.roll2 = roll2;
    this.roll3 = roll3;
    this.score = score;
    this.totalScore = totalScore;
  }

  static fromStringObj(s: FrameValuesString): FrameValuesTyped {
    return new FrameValuesTyped(
      s.frame !== "" ? Number(s.frame) : 0,
      s.roll1,
      s.roll2,
      s.roll3,
      s.score,
      s.totalScore
    );
  }

  toStringObj(): FrameValuesString {
    return new FrameValuesString(
      String(this.frame),
      String(this.roll1),
      String(this.roll2),
      String(this.roll3),
      String(this.score),
      String(this.totalScore)
    );
  }

  static toStringList(list: FrameValuesTyped[]): FrameValuesString[] {
    return list.map(t => t.toStringObj());
  }

  static fromStringList(list: FrameValuesString[]): FrameValuesTyped[] {
    return list.map(s => FrameValuesTyped.fromStringObj(s));
  }

  toJsonValue(): Record<string, unknown> {
    return {
      frame: this.frame,
      roll1: this.roll1,
      roll2: this.roll2,
      roll3: this.roll3,
      score: this.score,
      totalScore: this.totalScore,
    };
  }

  toJSON(): string { return _json.stringify(this.toJsonValue()); }

  static fromJsonValue(m: unknown): FrameValuesTyped {
    return new FrameValuesTyped(
      _json.asInt(_json.requireField(m, "frame"), "frame"),
      _json.asString(_json.requireField(m, "roll1"), "roll1"),
      _json.asString(_json.requireField(m, "roll2"), "roll2"),
      _json.asString(_json.requireField(m, "roll3"), "roll3"),
      _json.asString(_json.requireField(m, "score"), "score"),
      _json.asString(_json.requireField(m, "totalScore"), "totalScore")
    );
  }

  static fromJSON(text: string): FrameValuesTyped {
    return FrameValuesTyped.fromJsonValue(_json.parse(text));
  }

  static toJSONList(items: readonly FrameValuesTyped[]): string {
    return _json.stringify(items.map((item) => item.toJsonValue()));
  }

  static fromJSONList(text: string): FrameValuesTyped[] {
    const raw = _json.asArray(_json.parse(text), "FrameValuesTyped") ?? [];
    return raw.map((e) => FrameValuesTyped.fromJsonValue(e));
  }

  toString(): string {
    return `Frame=${this.frame}, Roll1=${this.roll1}, Roll2=${this.roll2}, Roll3=${this.roll3}, Score=${this.score}, TotalScore=${this.totalScore}`;
  }

  equals(other: FrameValuesTyped): boolean {
    return this.frame === other.frame
      && this.roll1 === other.roll1
      && this.roll2 === other.roll2
      && this.roll3 === other.roll3
      && this.score === other.score
      && this.totalScore === other.totalScore;
  }
}

File: Python/ExampleTests/common/frame_values_typed.py
from . import json_util as _json
from .frame_values_string import FrameValuesString

class FrameValuesTyped:
    def __init__(self, frame: int = 0, roll1: str = '', roll2: str = '', roll3: str = '', score: str = '', total_score: str = ''):
        self.frame = frame
        self.roll1 = roll1
        self.roll2 = roll2
        self.roll3 = roll3
        self.score = score
        self.total_score = total_score

    @classmethod
    def from_string_obj(cls, s: FrameValuesString) -> 'FrameValuesTyped':
        return cls(
            int(s.frame) if s.frame else 0,
            s.roll1,
            s.roll2,
            s.roll3,
            s.score,
            s.total_score
        )

    def to_string_obj(self) -> FrameValuesString:
        return FrameValuesString(
            str(self.frame),
            str(self.roll1),
            str(self.roll2),
            str(self.roll3),
            str(self.score),
            str(self.total_score)
        )

    @staticmethod
    def to_string_list(items) -> list:
        return [t.to_string_obj() for t in items]

    @staticmethod
    def from_string_list(items) -> list:
        return [FrameValuesTyped.from_string_obj(s) for s in items]

    def to_json_value(self) -> dict:
        return {
            'Frame': self.frame,
            'Roll1': self.roll1,
            'Roll2': self.roll2,
            'Roll3': self.roll3,
            'Score': self.score,
            'TotalScore': self.total_score,
        }

    def to_json(self) -> str:
        return _json.dumps(self.to_json_value())

    @classmethod
    def from_json_value(cls, m: dict) -> 'FrameValuesTyped':
        return cls(
            _json.as_int(_json.require(m, 'Frame'), 'Frame'),
            _json.as_str(_json.require(m, 'Roll1'), 'Roll1'),
            _json.as_str(_json.require(m, 'Roll2'), 'Roll2'),
            _json.as_str(_json.require(m, 'Roll3'), 'Roll3'),
            _json.as_str(_json.require(m, 'Score'), 'Score'),
            _json.as_str(_json.require(m, 'TotalScore'), 'TotalScore')
        )

    @classmethod
    def from_json(cls, text: str) -> 'FrameValuesTyped':
        return cls.from_json_value(_json.loads(text))

    @staticmethod
    def to_json_list(items) -> str:
        return _json.dumps([item.to_json_value() for item in items])

    @classmethod
    def from_json_list(cls, text: str) -> list:
        raw = _json.as_list(_json.loads(text), 'FrameValuesTyped')
        return [cls.from_json_value(e) for e in raw]

    def __str__(self):
        return (f'Frame={self.frame}' + ', ' +
                f'Roll1={self.roll1}' + ', ' +
                f'Roll2={self.roll2}' + ', ' +
                f'Roll3={self.roll3}' + ', ' +
                f'Score={self.score}' + ', ' +
                f'TotalScore={self.total_score}')

    def _key(self):
        return (self.frame, self.roll1, self.roll2, self.roll3, self.score, self.total_score)

    def __eq__(self, other):
        if not isinstance(other, FrameValuesTyped):
            return NotImplemented
        return self._key() == other._key()

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'FrameValuesTyped({self})'
